src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -972,15 +972,21 @@ private static int CanonicallyCompareDiagnostics(Diagnostic x, Diagnostic y)
 
             int codeCompare = xCode.CompareTo(yCode);
 
-            // ToString fails for a diagnostic with an error code that does not prevent successful delegate conversion.
```

Comment:
>ToString fails for a diagnostic with an error code that does not prevent successful delegate conversion [](start = 15, length = 103)

Where was the failure previously? In converting the entire diagnostic to a string or converting individual arguments?

Reply:
Hmmm.

I can't say, as I can't reproduce that exception. ToString works just fine on the cases I've tested.

I also can't see any reason why an exception should occur.

It's difficult to investigate further, since the comment goes back to the original git commit of roslyn.

Perhaps the safest thing to do is to revert my change and a comment explaining that there may or may not be an exception thrown if you remove this code?

Reply:
actually I think I know what it might be - if the error code is Unknown or void, then calling toString will throw.

Reply:
Sometimes the arguments are Symbols for which ToString is a complex operation, which for all I know may sometimes throw an exception. It's also more expensive than calling ErrorFacts.PreventsSuccessfulDelegateConversion().

So to be safe I've decided to call that, and explained why in a comment.

In the process I realized that Diagnostics.Arguments can be null, so I've checked for that case.

Reply:
Actually @jmarolf found the original commit:

> Here is the original commit message by Andrew Casey:
>
> UnboundLambdaState.GenerateSummaryErrors uses string.CompareOrdinal to determine whether two diagnostics are the same. This does not work in the case where one (or both) of the diagnostics is Void (part of the obsolete mechanism), because such diagnostics do not have messages. In such cases, simply compare the error codes.
While we're at it, we may as well compare the codes as an acid test in the general case too

In which case it's fine to omit the check, as Void diagnostics have no arguments

Reply:
Here is the test case originally effected by this change:
https://github.com/dotnet/roslyn/blob/e5b79f84fc66d63db1ad4e194b38655fc1cdbbf9/src/Compilers/CSharp/Test/Symbol/Compilation/GetSemanticInfoBrokenCodeTests.cs#L135

Reply:
That test now passes even if we call ToString() on the method